bpf/lib/arp.h

```diff
@@ -29,54 +29,62 @@ static __always_inline int arp_check(struct ethhdr *eth, struct arphdr *arp,
 }
 
 static __always_inline int
-arp_prepare_response(struct __ctx_buff *ctx, struct ethhdr *eth,
-		     struct arp_eth *arp_eth, __be32 ip, union macaddr *mac)
+arp_prepare_response(struct __ctx_buff *ctx, union macaddr *smac, __be32 sip,
+		     union macaddr *dmac, __be32 tip)
 {
-	union macaddr smac = *(union macaddr *) &eth->h_source;
-	__be32 sip = arp_eth->ar_sip;
 	__be16 arpop = bpf_htons(ARPOP_REPLY);
 
-	if (eth_store_saddr(ctx, mac->addr, 0) < 0 ||
-	    eth_store_daddr(ctx, smac.addr, 0) < 0 ||
+	if (eth_store_saddr(ctx, smac->addr, 0) < 0 ||
+	    eth_store_daddr(ctx, dmac->addr, 0) < 0 ||
 	    ctx_store_bytes(ctx, 20, &arpop, sizeof(arpop), 0) < 0 ||
-	    ctx_store_bytes(ctx, 22, mac, 6, 0) < 0 ||
-	    ctx_store_bytes(ctx, 28, &ip, 4, 0) < 0 ||
-	    ctx_store_bytes(ctx, 32, &smac, sizeof(smac), 0) < 0 ||
-	    ctx_store_bytes(ctx, 38, &sip, sizeof(sip), 0) < 0)
+	    ctx_store_bytes(ctx, 22, smac, 6, 0) < 0 ||
+	    ctx_store_bytes(ctx, 28, &sip, 4, 0) < 0 ||
+	    ctx_store_bytes(ctx, 32, dmac, 6, 0) < 0 ||
+	    ctx_store_bytes(ctx, 38, &tip, 4, 0) < 0)
```

Comment:
nit: Is there a reason to avoid `sizeof()` here? 

```suggestion
	    ctx_store_bytes(ctx, 22, smac, sizeof(*smac), 0) < 0 ||
	    ctx_store_bytes(ctx, 28, &sip, sizeof(sip), 0) < 0 ||
	    ctx_store_bytes(ctx, 32, dmac, sizeof(*dmac), 0) < 0 ||
	    ctx_store_bytes(ctx, 38, &tip, sizeof(tip), 0) < 0)
```

Reply:
It looks like bpf verifier does not like `sizeof(*smac)` at all. Might have to do with how static data is being defined? Keeping a literal there for now as it was in the original version.